Feature/reset password route

agusbusta opened the pull request:
This pull request introduces enhancements to the user authentication flow by integrating Auth0 for password management and implementing an email service for user notifications. The changes primarily focus on the forgot_password and reset_password routes, allowing users to securely reset their passwords via email.

Changes Made

1. Auth0 Integration
Implemented the patchPassword function to interact with the Auth0 Management API. This function retrieves the user ID based on the provided email and updates the user's password securely.
Utilized JWT (JSON Web Tokens) to generate a unique token for password reset requests, ensuring that the reset link is valid for a limited time (1 hour).
2. Email Service Implementation
Created an EmailService class to encapsulate the email sending functionality using Flask-Mail. This service includes methods for sending registration confirmation emails and password reset emails.
Integrated the send_password_reset_email method into the forgot_password route. When a user requests a password reset, an email containing a reset link is sent to the user's registered email address.
The email includes a personalized message and a link that directs the user to the password reset page.
Route Modifications

forgot_password route:
Updated to validate the user's email and send a password reset email using the new email service.
reset_password route:
Enhanced to decode the JWT token, validate it, and update the user's password through the Auth0 API.

daviddflix commented on routes/user/auth.py:
```diff
@@ -0,0 +1,62 @@
+
+
+async def get_management_api_token():
+    url = f"https://{auth0Domain}/oauth/token"
```
Here, we need to handle errors with a try/except block in case the token is not generated.

daviddflix commented on routes/user/auth.py:
```diff
@@ -0,0 +1,62 @@
+    data = response.json()
+    return data['access_token']
+
+async def patchPassword(email, new_password):
```
This function is good, however, it's not returning anything that lets the parent function know if the password was updated. for this I suggest

- Raising the exception here, which the parent function should catch and handle.
- Please remove all prints after the testing.

daviddflix commented on routes/user/user.py:
```diff
@@ -526,3 +537,91 @@ def delete_user_account():
+    data = request.json
+
+    email = data.get('email')
+    if not email or not validate_email(email):  # Implement validate_email function
```
Please remove the comment!

daviddflix commented on routes/user/user.py:
```diff
@@ -526,3 +537,91 @@ def delete_user_account():
+
+    try:
+        # Decode the token
+        payload = jwt.decode(token, current_app.config['SECRET_KEY'], algorithms=['HS256'])
```
There is an important step missing here. After decoding the token, we must verify its validity before proceeding to update the password.

Please refer to the ADMIN table, which includes methods specifically designed for this purpose. We should do the same for the USER table. Let me know if you need any assistance or clarification while implementing this.

agusbusta commented:
ok changes are done but still need to test when Rajan provides OPT CODE.

agusbusta commented:
### feat(auth): Improve password reset functionality and security

#### Summary
This PR enhances the password reset functionality by improving security, error handling, and code quality.

---

#### Changes:

##### 1. Password Strength Validation
- Enforce specific requirements:
  - Minimum 8 characters
  - At least one uppercase letter
  - At least one lowercase letter
  - At least one number
  - At least one special character

##### 2. Enhanced Error Handling for Auth0 Password Updates
- Better detection of authentication providers (Google, Apple)
- Improved validation of user connection types
- Clear error messages for users

##### 3. Code Quality Improvements
- Add comprehensive docstrings
- Remove debug prints
- Add type hints
- Enhance code readability
- Optimize error messages

##### 4. Password Reset Template Updates
- Add password requirements display
- Improve error message presentation
- Add validation feedback



Now it's ready to review!

daviddflix reviewed (changes requested):
**Suggested Improvements:**

1. **Async HTTP Requests:**  
   The `reset_password` route is asynchronous, but the requests to Auth0 are made using the `requests` library, which is blocking. Consider using an asynchronous HTTP client library (e.g., `httpx` or `aiohttp`) to prevent blocking the event loop.

2. **Consistent Time Handling for JWT:**  
   For token expiration, consider using `datetime.datetime.now()` instead of `datetime.datetime.utcnow()`. which is depracated.

5. **Use Structured Logging Instead of Prints:**  
   Replace `print` statements in `patchPassword` with proper logging for better observability and to avoid leaking sensitive information in logs.

6. **Token Invalidation and Storage (IMPORTANT):**  
   Consider storing the reset token in the database and marking it as used once the password is reset. This prevents tokens from being reused and adds an extra layer of security to the reset flow.

agusbusta commented:
Hey @daviddflix  the changes are ready! also I resolve some conflicts from develop branch merging into this branch!

daviddflix reviewed (changes requested):
Hey @agusbusta everything look great, however, there are two things we should address.

1. The reset endpoint is using a verify token method, which should be defined in the User class.

2. The auth method are using async, however when making the request the function uses an sync library "Request", we should either use httpx an async library or change the function definition.

Any question let me know

agusbusta commented:
Recent changes includes two main improvements to our authentication system:

1. Replaces the synchronous `requests` library with `httpx` to properly handle async operations in our Auth0 integration. This ensures consistency with our async/await pattern and improves performance for concurrent requests.

2. Adds token verification functionality to the User class, which was missing but required for the password reset feature to work properly. This method securely validates reset tokens before allowing password changes.

Technical Changes:
- Migrated auth operations to use async HTTP client (httpx)
- Added verify_reset_token method to User model
- Updated error handling for async operations
- Ensured proper JWT validation for password reset flow

daviddflix commented:
Pull Request Feedback

Hey @agusbusta,

The code looks much better now! Great work! However, there are a couple of changes I'd like to propose:

A new Table should be created to store the user token my suggestion is:
python
```
class UserPasswordReset(Base):
    __tablename__ = 'password_resets'
    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey('users.id'))
    reset_code = Column(String(6), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    expires_at = Column(DateTime, nullable=False)
    is_used = Column(Boolean, default=False)
```

**Token Exposure:** The forgot password endpoint is sending the raw token, this is a security risk and should be avoided, please consider other alternatives.

**Suggested dummy code**

python
```
@user_bp.route('/forgot-password', methods=['POST'])
def forgot_password():
    """
    Request a password reset token for the user.
    """
    response = {'success': False, 'message': None}
    data = request.json
    
    email = data.get('email')
    if not email or not validate_email(email):
        response['message'] = 'Invalid email format'
        return jsonify(response), 400
    
    with Session() as session:
        try:
            user = session.query(User).filter_by(email=email).first()
            if not user:
                # Use same message as success for security
                response['message'] = 'If your email is registered, you will receive reset instructions'
                return jsonify(response), 200
            
            # Generate reset token with shorter expiration (15 minutes)
            token = user.generate_reset_token(expires_in=900)
            
            # Instead of including token in URL, send a reset code page URL
            dashboard_url = 'https://dashboard-alpha-ai.vercel.app/#/reset-password'
            
            # Send password reset email with token
            email_service.send_password_reset_email(
                user.email,
                user.nickname,
                token,
                dashboard_url
            )
            
            response['success'] = True
            response['message'] = 'If your email is registered, you will receive reset instructions'
            return jsonify(response), 200
            
        except Exception as e:
            logging.error(f"Password reset error for email {email}: {str(e)}")
            response['message'] = 'An error occurred processing your request'
            return jsonify(response), 500
```

**Suggested email service method:**

python
```
def send_password_reset_email(self, user_email, username, token, dashboard_url):
    """Send password reset email with token."""
    self.logger.info(f"Sending password reset email to {user_email}")
    subject = "Password Reset Request - AI Alpha Dashboard"
    
    # Create a state parameter with the token
    state = base64.urlsafe_b64encode(token.encode()).decode()
    
    html_body = render_template(
        'password_reset_email.html',
        username=username,
        dashboard_url=f"{dashboard_url}?state={state}",
        expiry_minutes=15
    )
    
    try:
        self.send_email(to=user_email, subject=subject, body="", html=html_body)
        self.logger.info(f"Password reset email sent successfully to {user_email}")
    except Exception as e:
        self.logger.error(f"Failed to send password reset email to {user_email}: {str(e)}")
        raise
```

**API Documentation**: We need to update the Swagger documentation to include the two new endpoints. Make sure to update this branch accordingly, as we are using a single file for the entire documentation. Since you'll be introducing new lines, be mindful of potential conflicts with other changes that might be occupying the same lines.

**Logging in Docker Environment:** As you know, using print statements is not ideal in a Docker environment. The email service in the code is much more organized and uses a proper logger. I recommend adopting this approach for logging in your code as well. This will ensure better organization and avoid potential issues in production.

python
```
import os
import logging
from flask import current_app, render_template
from flask_mail import Message, Mail
from dotenv import load_dotenv

# Force reload environment variables
load_dotenv(override=True)

class EmailService:
    def __init__(self, app=None):
        self.mail = None
        self.logger = logging.getLogger(__name__)
        self._setup_logger()
        if app:
            self.init_app(app)

    def _setup_logger(self):
        """Configure the logger with appropriate formatting and handlers"""
        formatter = logging.Formatter(
            '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
        
        # Console handler
        console_handler = logging.StreamHandler()
        console_handler.setFormatter(formatter)
        self.logger.addHandler(console_handler)
        
        # File handler
        file_handler = logging.FileHandler('email_service.log')
        file_handler.setFormatter(formatter)
        self.logger.addHandler(file_handler)
        
        self.logger.setLevel(logging.INFO)

    def init_app(self, app):
        # Get email configuration from environment
        username = str(os.environ.get('MAIL_USERNAME', '')).strip()
        password = str(os.environ.get('MAIL_PASSWORD', '')).strip()
        sender = str(os.environ.get('MAIL_DEFAULT_SENDER', '')).strip()

        self.logger.info(f"Initializing EmailService with username: {username}, sender: {sender}")
        
        app.config.update(
            MAIL_SERVER='smtp.gmail.com',
            MAIL_PORT=465,
            MAIL_USE_TLS=False,
            MAIL_USE_SSL=True,
            MAIL_USERNAME=username,
            MAIL_PASSWORD=password,
            MAIL_DEFAULT_SENDER=sender
        )
        
        self.mail = Mail(app)
        self.logger.info("Email service initialized successfully")

    def send_email(self, to, subject, body, html=None): 
        if not self.mail:
            self.logger.warning("Mail instance not found, initializing with current_app")
            self.mail = Mail(current_app)
        
        try:
            self.logger.info(f"Attempting to send email to: {to}")
            sender = current_app.config.get('MAIL_DEFAULT_SENDER')
            msg = Message(subject, sender=sender, recipients=[to], body=body, html=html)
            self.mail.send(msg)
            self.logger.info(f"Email sent successfully to {to}")
        except Exception as e: 
            error_msg = f'Error sending email: {str(e)}'
            self.logger.error("="*50)
            self.logger.error("Email Error Details:")
            self.logger.error(error_msg)
            self.logger.error("="*50)
            raise Exception(error_msg)
     
    def send_registration_confirmation(self, to, username, temporary_password):
        self.logger.info(f"Sending registration confirmation to {to} for username: {username}")
        subject = "Welcome to AI Alpha Dashboard"
        dashboard_url = 'https://dashboard-alpha-ai.vercel.app/#/login'  

        html_body = render_template('welcome_email.html', 
                                    username=username, 
                                    temporary_password=temporary_password, 
                                    dashboard_url=dashboard_url)
        
        try:
            self.send_email(to, subject, body="", html=html_body)
            self.logger.info(f"Registration confirmation sent successfully to {to}")
        except Exception as e:
            self.logger.error(f"Failed to send registration confirmation to {to}: {str(e)}")
            raise

    def send_password_reset_email(self, user_email, username, reset_link):
        self.logger.info(f"Sending password reset email to {user_email} for username: {username}")
        subject = "Password Reset Request - AI Alpha Dashboard"
        html_body = render_template('password_reset_email.html',
                            username=username,
                            reset_link=reset_link)
        
        try:
            self.send_email(to=user_email, subject=subject, body="", html=html_body)
            self.logger.info(f"Password reset email sent successfully to {user_email}")
        except Exception as e:
            self.logger.error(f"Failed to send password reset email to {user_email}: {str(e)}")
            raise
```
